File: code/758/solution.py
mod = int(1e9 + 7)


def add(a, b):
    return (a + b) % mod


def mul(a, b):
    return a * b % mod


def qpow(a, b):
    ret = 1
    while b > 0:
        if (b & 1) != 0:
            ret = mul(ret, a)
        a = mul(a, a)
        b = int(b >> 1)
    return ret


def cal(a, b):
    if a < b:
        return 0
    x1 = qpow(2, a % b)
    x2 = qpow(qpow(2, b), int(a / b))
    x3 = add(1, mod - x2)
    x4 = qpow(add(1, mod - qpow(2, b)), mod - 2)
    return mul(x1, mul(x3, x4))


def exgcd(a, b):
    if b == 0:
        return 1, 0, True
    x, y, y_neg = exgcd(b, a % b)
    return y, add(x, mod - mul(y, cal(a, b))), not y_neg


def get_ans(a, b):
    x, y, y_neg = exgcd(a, b)
    x = mul(x, 2)
    y = mul(y, 2)
    if y_neg:
        y = add(mod, -y)
    else:
        x = add(mod, -x)
    return add(add(mod, -2), add(x, y))


def is_prime(a):
    for i in range(2, a):
        if a % i == 0:
            return False
    return True


if __name__ == "__main__":
    ans = 0
    prime = []
    for i in range(2, 1000):
        if is_prime(i):
            prime.append(i * i * i * i * i)
    for i in range(len(prime)):
        for j in range(i + 1, len(prime)):
            ans = add(ans, get_ans(prime[i], prime[j]))
    print(ans)

File: code/758/solution.cpp
#include <iostream>
#include <vector>
using namespace std;
const int mod = 1e9 + 7;
int add(int a, int b) { return (a + b) % mod; }
int mul(int a, int b) { return 1ll * a * b % mod; }
int qpow(int a, long long b) {
    int ret = 1;
    while (b) {
        if (b & 1)
            ret = mul(ret, a);
        a = mul(a, a);
        b >>= 1;
    }
    return ret;
}
int cal(long long a, long long b) {
    // return (2^a-1) / (2^b-1)
    if (a < b)
        return 0;
    auto x1 = qpow(2, a % b);
    auto x2 = qpow(qpow(2, b), a / b);
    auto x3 = add(1, mod - x2);
    auto x4 = qpow(add(1, mod - qpow(2, b)), mod - 2);
    return mul(x1, mul(x3, x4));
}
bool exgcd(long long a, long long b, int &x, int &y) {
    if (b == 0) {
        x = 1;
        y = 0;
        return true;
    }
    auto y_neg = exgcd(b, a % b, x, y);
    int t      = y;
    y          = add(x, mod - mul(y, cal(a, b)));
    x          = t;
    return !y_neg;
}
int get_ans(long long a, long long b) {
    int x, y;
    bool y_neg = exgcd(a, b, x, y);
    x          = mul(x, 2);
    y          = mul(y, 2);
    if (y_neg) {
        y = add(mod, -y);
    } else {
        x = add(mod, -x);
    }
    return add(add(mod, -2), add(x, y));
}
bool is_prime(int a) {
    for (int i = 2; i < a; i++)
        if (a % i == 0)
            return false;
    return true;
}

int main() {
    int ans = 0;
    vector<long long> prime;
    for (int i = 2; i < 1000; i++)
        if (is_prime(i))
            prime.push_back(1ll * i * i * i * i * i);
    for (int i = 0; i < prime.size(); i++) {
        for (int j = i + 1; j < prime.size(); j++) {
            ans = add(ans, get_ans(prime[i], prime[j]));
        }
    }
    cout << ans << endl;
    return 0;
}
